### PR TITLE
PEP 545

### PR DESCRIPTION
**Pull request by JulienPalard**
This implements https://www.python.org/dev/peps/pep-0545/, building all known translations by default (only french for the moment, japanese may arrive soon.

This implements *no* crosslink between versions (no language switcher in the graphical interface) allowing to test for a period of time before going really public about it.

I tested it with:

```
$ mkdir /tmp/build_root  /tmp/www_root /tmp/log_root
$ git clone https://github.com/python/cpython.git --branch 3.6 --depth 1 /tmp/build_root/python36
$ git clone https://github.com/python/cpython.git --branch 2.7 --depth 1 /tmp/build_root/python27
$ git clone https://github.com/python/cpython.git --branch 3.5 --depth 1 /tmp/build_root/python35
$ git clone https://github.com/python/cpython.git --branch master --depth 1 /tmp/build_root/python37
$ python3 -m venv /tmp/build_root/environment/
$ . /tmp/build_root/environment/bin/activate
$ python3 -m pip install -r requirements.txt
$ python3 ./build_docs.py --git --build-root /tmp/build_root/ --www-root /tmp/www_root/ --log-directory /tmp/log_root/ --skip-cache-invalidation --group $(id -gn)
```

And it went well, properly skipping untranslated versions, building a correct hierarchy in www_root. It'll probably miss some work before implementing cross-linking like setting symlinks like /3/, and /dev/, which are currently done manually if I'm right, that I should implement in salt or in this script in a near future.

### CODE REVIEW

**Comment by JulienPalard** [comment]
highlighting @ned-deily @ewdurbin @benjaminp for review.

File: build_docs.py [hunk 1]
```diff
@@ -80,11 +95,57 @@ def changed_files(directory, other):
     return changed
 
 
+def git_clone(repository, directory, branch='master'):
+    """Clone or update the given branch of the given repository in the
+    given directory.
+    """
+    logging.info("Updating repository %s in %s", repository, directory)
+    try:
+        shell_out("git -C {} checkout {}".format(directory, branch))
```
Comment (haypo) [inline build_docs.py]: Would it be possible to pass a list of arguments rather than a string to avoid issues with shell?

File: build_docs.py [hunk 2]
```diff
@@ -98,8 +159,9 @@ def build_one(version, isdev, quick, sphinxbuild, build_root, www_root,
     maketarget = "autobuild-" + ("dev" if isdev else "stable") + ("-html" if quick else "")
     logging.info("Running make %s", maketarget)
     logname = os.path.basename(checkout) + ".log"
-    shell_out("cd Doc; make SPHINXBUILD=%s %s >> %s 2>&1" %
-              (sphinxbuild, maketarget, os.path.join(log_directory, logname)))
+    shell_out("cd Doc; make SPHINXBUILD=%s SPHINXOPTS='%s' %s >> %s 2>&1" %
+              (sphinxbuild, sphinxopts, maketarget,
+               os.path.join(log_directory, logname)))
```
Comment (haypo) [inline build_docs.py]: Can't use use cwd parameter of subprocess.Popen to avoid the "cd"?

Note: make also has a -C parameter to make the "cd".

Same question for string vs list.

File: build_docs.py (hunk 1 above)
Comment (haypo) [inline build_docs.py]: Oh, wait, the "{}" or "%s" in shell commands are not new: the current code already use them. You can ignore my comment. It can be enhanced in a different PR ;-)

File: build_docs.py (hunk 2 above)
Comment (JulienPalard) [inline build_docs.py]: That's right, it will even allow running make directly (without a shell) once your previous comment fixed, which is nice. I tried to keep my changes minimal so I didn't changed this. Will do in another PR, later.

(As I don't have access to the logs or the server, I prefer keeping my chances atomic, so if something break we all know exactly what caused it).

File: build_docs.py (hunk 1 above)
Comment (JulienPalard) [inline build_docs.py]: Will gladly fix this in another PR yes.

File: build_docs.py (hunk 2 above)
Comment (haypo) [inline build_docs.py]: Ok, it perfectly makes sense. But it would help if someone could give you access to the logs since you seem to be one of the people working on this project :-)

@ewdurbin: Can you please give access to Julien to the logs? Or maybe send him logs on demand? Thanks.

FYI I proposed to give the commit bit to @JulienPalard on this project :-)
https://mail.python.org/pipermail/python-committers/2017-June/004583.html

Comment (MarkMangoba) [inline build_docs.py]: @haypo I can help get the logs, can you point me to what logs you'll need?

Comment (JulienPalard) [inline build_docs.py]: @MarkMangoba @haypo is speaking about the output of this crontab: https://github.com/python/psf-salt/blob/master/salt/docs/init.sls#L47 and [/var/log/docsbuild/](https://github.com/python/docsbuild-scripts/blob/master/build_docs.py#L205).

Comment (MarkMangoba) [inline build_docs.py]: @JulienPalard understood.  @ewdurbin do you have a chance to review?

**Comment by haypo** [comment]
@MarkMangoba, @ewdurbin & others: how can we unblock this situation please?

**Comment by ewdurbin** [comment]
Running test builds now! Thanks @JulienPalard

**Comment by haypo** [comment]
Cool, thanks @ewdurbin! How can we check for failure? Are you going to notify us? @JulienPalard: maybe join the #python-infra channel on Freenode? ;-)

**Comment by ewdurbin** [comment]
@haypo indeed, unfortunately it's just logs on the docs server when the build job kicks off via cron.
